agent/internal/container/container.go

```diff
@@ -0,0 +1,428 @@
+	}
+	killed.Store(false)
+
+	c.log.Trace("running container")
+	dc, err := c.docker.RunContainer(ctx, dockerID)
+	if err != nil {
+		return err
+	}
+	if err := c.running(ctx, aproto.ContainerStarted{ContainerInfo: dc.ContainerInfo}); err != nil {
+		return err
+	}
+
+	return c.wait(ctx, dc)
+}
+
+func (c *Container) reattach(ctx context.Context) error {
+	c.log.Trace("entering reattach")
+	switch dc, exitCode, err := c.docker.ReattachContainer(
+		ctx,
+		docker.LabelFilter(docker.ContainerIDLabel, c.containerID.String()),
+	); {
+	case errors.Is(err, context.Canceled):
+		return err
+	case err != nil:
+		return aproto.NewContainerFailure(aproto.RestoreError, err)
+	case exitCode != nil:
+		return aproto.NewContainerExit(*exitCode)
+	case dc == nil:
+		return aproto.NewContainerFailure(aproto.RestoreError, errors.New("container unknown to docker"))
+	default:
+		return c.wait(ctx, dc)
+	}
+}
+
+func (c *Container) wait(ctx context.Context, dc *docker.Container) error {
+	c.log.Trace("in monitoring loop")
+	for {
+		select {
+		case exit := <-dc.ContainerWaiter.Waiter:
+			c.log.Trace("container exited")
+			if exit.Error != nil {
+				return fmt.Errorf("receiving container exit: %s", exit.Error.Message)
+			}
+			return aproto.NewContainerExit(aproto.ExitCode(exit.StatusCode))
+
+		case err := <-dc.ContainerWaiter.Errs:
+			c.log.Trace("container waiter failed")
+			return fmt.Errorf("failed while waiting for container to exit: %w", err)
+
+		case signal := <-c.signals:
+			c.log.Tracef("container signaled: %s", signal)
+			if err := c.docker.SignalContainer(ctx, dc.ContainerInfo.ID, signal); err != nil {
+				c.log.WithError(err).Errorf(
+					"failed to signal %v with %v", dc.ContainerInfo.ID, signal,
+				)
+			}
+
+		case <-ctx.Done():
+			c.log.Trace("container context canceled")
+			return ctx.Err()
+		}
+	}
+}
+
+func (c *Container) finalize(ctx context.Context, err error) {
+	c.log.Trace("finalizing container exit")
+	if ctx.Err() != nil {
+		// There is a chance that cancellation and some other error raced, meaning we have a
+		// valid error and a canceled context. In this case, we just go ahead with the detach
+		// flow - on reattach callers can just reinspect the container.
+		c.log.
+			WithError(err).
+			WithField("ctx-err", ctx.Err()).
+			Warnf("orphaning container")
+		return
+	}
+
+	var stop aproto.ContainerStopped
+	switch err := err.(type) {
+	case nil:
+		stop = aproto.ContainerStopped{Failure: nil}
+	case *aproto.ContainerFailure:
+		stop = aproto.ContainerStopped{Failure: err}
+	default:
+		stop = aproto.ContainerError(aproto.TaskError, err)
+	}
+
+	if err := c.terminated(ctx, stop); err != nil {
+		c.log.WithError(err).Error("finalizing container")
+	}
+	return
+}
+
+func (c *Container) summary() cproto.Container {
+	devices := make([]device.Device, len(c.devices))
+	copy(devices, c.devices)
+	return cproto.Container{
+		ID:      c.containerID,
+		State:   c.state,
+		Devices: devices,
+	}
```

Comment:
every time I use `copy` I end up doing something wrong
```suggestion
	return cproto.Container{
		ID:      c.containerID,
		State:   c.state,
		Devices: slices.Clone(c.devices),
	}
```